pkg/storagespace/storagespace.go

```diff
@@ -0,0 +1,151 @@
+// Copyright 2018-2021 CERN
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// In applying this license, CERN does not waive the privileges and immunities
+// granted to it by virtue of its status as an Intergovernmental Organization
+// or submit itself to any jurisdiction.
+
+package storagespace
+
+import (
+	"path"
+	"strings"
+
+	provider "github.com/cs3org/go-cs3apis/cs3/storage/provider/v1beta1"
+	"github.com/cs3org/reva/v2/pkg/utils"
+	"github.com/pkg/errors"
+)
+
+const (
+	_idDelimiter        string = "!"
+	_storageIDDelimiter string = "$"
+)
+
+var (
+	// ErrInvalidSpaceReference signals that the space reference is invalid.
+	ErrInvalidSpaceReference = errors.New("invalid storage space reference")
+	// ErrInvalidSpaceID signals that the space ID is invalid.
+	ErrInvalidSpaceID = errors.New("invalid storage space id")
+)
+
+// SplitID splits a storage space ID into a provider ID and a node ID.
+// The accepted formats results of the storage space ID and respective results
+// are:
+// <providerid>$<spaceid>!<nodeid> 	-> <providerid>$<spaceid>, <nodeid>
+// <providerid>$<spaceid>			-> <providerid>$<spaceid>, <spaceid>
+// <spaceid>						-> <spaceid>, <spaceid>
+func SplitID(ssid string) (storageid, nodeid string, err error) {
+	if ssid == "" {
+		return "", "", errors.Wrap(ErrInvalidSpaceID, "can't split empty storage space ID")
+	}
+	parts := strings.SplitN(ssid, _idDelimiter, 2)
+	if len(parts) == 1 || parts[1] == "" {
+		_, sid := SplitStorageID(parts[0])
+		return parts[0], sid, nil
+	}
+	return parts[0], parts[1], nil
+}
+
+// SplitStorageID splits a storage ID into the provider ID and the spaceID.
+// The accepted formats are:
+// <providerid>$<spaceid>			-> <providerid>, <spaceid>
+// <spaceid>						-> "", <spaceid>
+func SplitStorageID(sid string) (providerID, spaceID string) {
+	parts := strings.SplitN(sid, _storageIDDelimiter, 2)
+	if len(parts) == 2 {
+		return parts[0], parts[1]
+	}
+	return "", parts[0]
+}
+
+// FormatResourceID converts a ResourceId into the string format.
+// The result format will look like:
+// <storageid>!<opaqueid>
+func FormatResourceID(sid provider.ResourceId) string {
```

Comment:
Could this take `*provider.ResourceId` to avoid having to dereference the `sid` all the time? We only ever work with `*ResourceId` usually

Reply:
I really dislike using pointers everywhere. IMO the fact that the ResourceId is a pointer almost everywhere else is a bad design so I'd like to slowly reduce that but if you insist I can change that to use pointers.

Reply:
I understand. But the question is: Is there a realistic chance that cs3apis change from using pointers to absolute values? If not we will have do dereferencing and referencing all the time. This will lead to a worse performance and ultimately to more bugs.

So long story short: I don't care too much if it is a pointer or a struct, but I would prefer to have it consistent. Afais cs3api won't change at any time to use structs instead of pointers. But if you tell me they will, I will believe you...

Reply:
> Is there a realistic chance that cs3apis change from using pointers to absolute values?

I see your point. Probably not, at least not in the short-term.

> If not we will have do dereferencing and referencing all the time. This will lead to a worse performance and ultimately to more bugs.

Performance wise it should be a net zero. Yes, copying the memory for stack allocations do have some costs but the compiler can then better optimize those function call. Regarding bugs, I don't see how that would lead to bugs. Sure it's annoying to have to de-reference at many places but if you forget that your compiler will tell you so that shouldn't be too much of a problem. More an annoyance, I agree.

Just to write down  the result of our ~~offline~~ off-GitHub discussion:
We decided to leave it like that and try to reduce the use of pointers generally. It would also make sense to define a coding style guide we can refer to. Maybe something like: https://github.com/uber-go/guide/blob/master/style.md
But that is not part of this pull request.